src/site/antora/modules/ROOT/pages/manual/pattern-layout.adoc

```diff
@@ -0,0 +1,1719 @@
+characters, then truncate from the beginning.
+
+|%-20.-30c
+|true
+|20
+|30
+|Right pad with spaces if the category name is
+shorter than 20 characters. However, if the category name is longer than 30
+characters, then truncate from the end.
+|===
+
+[#jansi]
+=== ANSI styling on Windows
+
+ANSI escape sequences are supported natively on many platforms, but not by default on Windows.
+To enable ANSI support add the http://jansi.fusesource.org/[Jansi] dependency to your application, and set xref:manual/systemproperties.adoc#log4j.skipJansi[the `log4j.skipJansi` system property] to `false`.
+This allows Log4j to use Jansi to add ANSI escape codes when writing to the console.
+
+[NOTE]
+====
+Before Log4j 2.10, Jansi was enabled by default.
+The fact that Jansi requires native code means that Jansi can only be loaded by a single class loader.
+For web applications, this means the Jansi jar has to be in the web container's classpath.
+To avoid causing problems for web applications, Log4j no longer automatically tries to load Jansi without explicit configuration from Log4j 2.10 onward.
+====
+
+[#garbage-free]
+=== Garbage-free configuration
+
+Pattern Layout with the following limited set of conversion patterns is garbage-free.
+Format modifiers to control such things as field width, padding, left, and right justification will not generate garbage.
+
+[%header,cols="2m,2"]
+|===
+|Pattern
+|Comment
+
+|xref:#converter-logger[%c\{precision} +
+%logger\{precision}]
+|
+
+|xref:#converter-date[%d +
+%date]
+|Only the predefined date formats (`DEFAULT`, `ISO8601`, `UNIX`, `UNIX_MILLIS`, etc.) are garbage-free
+
+|xref:#converter-encode[%enc\{pattern} +
+%encode\{pattern}]
+|
+
+|xref:#converter-equals[%equals\{pattern}\{test}\{substitution} +
+%equalsIgnoreCase\{pattern}\{test}\{substitution}]
+|
+
+|xref:#converter-highlight[%highlight\{pattern}\{style}]
+|Granted nested pattern is garbage-free
+
+|xref:#converter-map[%K\{key} +
+%map\{key} +
+%MAP\{key}]
+|
+
+|xref:#converter-message[%m +
+%msg +
+%message]
+|Garbage-free unless message text contains `${`
+
+|xref:#converter-marker[%marker]
+|
+
+|xref:#converter-marker[%markerSimpleName]
+|
+
+|xref:#converter-max-len[%maxLen +
+%maxLength]
+|
+
+|%n
+|
+
+|xref:#converter-nano[%N +
+%nano]
+|
+
+|xref:#converter-not-empty[%notEmpty\{pattern} +
+%varsNotEmpty\{pattern} +
+%variablesNotEmpty\{pattern}]
+|
+
+|xref:#converter-level[%p +
+%level]
+|
+
+|xref:#converter-relative[%r +
+%relative]
+|
+
+|xref:#converter-seq[%sn +
+%sequenceNumber]
+|
+
+|xref:#converter-style[%style\{pattern}{ANSI style}]
+|
+
+|xref:#converter-thread-id[%T +
+%tid +
+%threadId]
+|
+
+|xref:#converter-thread-name[%t +
+%tn +
+%thread +
+%threadName]
+|
+
+|xref:#converter-thread-priority[%tp]
+|
+
+|xref:#converter-thread-context-map[%X{key[,key2...\]} +
+%mdc{key[,key2...\]} +
+%MDC{key[,key2...\]}]
+|
+
+|literal text
+|Garbage-free unless it contains `${`
```

Comment:
We should probably add a note about lazy substitution.

Adding `${k8s:containerId}` is garbage-free, because it will be expanded at configuration time. Adding `$${date:hh:mm:ss.SSS}` is not, since it will be expanded at each log event.

Reply:
Fixed in f77b5c0fce41b93da448f5d78fe902a09f970092.